x-pack/plugins/lens/public/shared_components/axis/extent/axis_extent_settings.tsx

```diff
@@ -228,6 +217,49 @@ function MetricAxisBoundsControl({
   );
 }
 
+export function getBounds(
+  mode: AxisExtentMode,
+  scaleType?: YScaleType,
+  dataBounds?: DataBoundsObject
+): Pick<UnifiedAxisExtentConfig, 'lowerBound' | 'upperBound'> {
+  if (mode !== 'custom' || !dataBounds) return {};
+
+  if (dataBounds.min >= 0 && dataBounds.max >= 0) {
+    const lowerBoundMax = scaleType === 'log' ? LOG_LOWER_BOUND_MAX : LOWER_BOUND_MAX;
+    return {
+      upperBound: dataBounds.max,
+      lowerBound: Math.min(lowerBoundMax, dataBounds.min),
+    };
+  }
+
+  if (dataBounds.min <= 0 && dataBounds.max <= 0) {
+    const upperBoundMin = scaleType === 'log' ? -LOG_LOWER_BOUND_MAX : LOWER_BOUND_MAX;
+    return {
+      upperBound: Math.max(upperBoundMin, dataBounds.min),
+      lowerBound: dataBounds.min,
+    };
+  }
+
+  if (scaleType === 'log') {
+    if (Math.abs(dataBounds.min) > Math.abs(dataBounds.max)) {
+      return {
+        upperBound: -LOG_LOWER_BOUND_MAX,
+        lowerBound: dataBounds.min,
+      };
+    }
+
+    return {
+      upperBound: dataBounds.max,
+      lowerBound: LOG_LOWER_BOUND_MAX,
+    };
+  }
+
+  return {
+    upperBound: dataBounds.max,
+    lowerBound: dataBounds.min,
+  };
+}
```

Comment:
This function gets the custom bounds given the scale constraints and domain range, changing the upper and lower bounds as needed.

This assumer a min value for log scales as `0.01` (aka `LOG_LOWER_BOUND_MAX`). This could be changed or configurable at some point.